zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs

```diff
@@ -0,0 +1,301 @@
+            from_bytes::expr(&a.cells[0..16]),
+            from_bytes::expr(&b.cells[0..16]),
+        );
+        let comparison_hi = ComparisonGadget::construct(
+            cb,
+            from_bytes::expr(&a.cells[16..32]),
+            from_bytes::expr(&b.cells[16..32]),
+        );
+        let a_lt_b_lo = lt_lo.expr();
+        let (a_lt_b_hi, a_eq_b_hi) = comparison_hi.expr();
+
+        // Add selector only for the cases where both a and b are positive or
+        // negative. This selector will be used after handling the cases
+        // where either only a or only b are negative.
+        //
+        // if (a > 0 && b > 0) || (a < 0 && b < 0):
+        //      a < b -> (a_hi < b_hi) ? 1 : (a_hi == b_hi) * (a_lo < b_lo)
+        //
+        // for e.g.: consider 8-bit signed integers -1 (0xff) and -2 (0xfe):
+        //     -2 < -1 and 0xfe < 0xff
+        let a_lt_b = select::expr(a_lt_b_hi, 1.expr(), a_eq_b_hi * a_lt_b_lo);
+
+        // Add a trivial selector: if only a or only b is negative we have the
+        // result.
+        // result = if a < 0 && b >= 0, slt = 1.
+        // result = if b < 0 && a >= 0, slt = 0.
+        let a_neg_b_pos = (1.expr() - a_pos.expr()) * b_pos.expr();
+        let b_neg_a_pos = (1.expr() - b_pos.expr()) * a_pos.expr();
+        let result = select::expr(
+            a_neg_b_pos,
+            1.expr(),
+            select::expr(b_neg_a_pos, 0.expr(), a_lt_b),
+        );
+
+        // Pop a and b from the stack, push the result on the stack.
+        cb.stack_pop(select::expr(is_sgt.expr(), b.expr(), a.expr()));
+        cb.stack_pop(select::expr(is_sgt.expr(), a.expr(), b.expr()));
+        cb.stack_push(result);
+
+        // The read-write counter changes by three since we're reading two words
+        // from stack and writing one. The program counter shifts only by one
+        // and the since the stack now has one less word, the stack pointer also
+        // shifts by one.
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(3.expr()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta(1.expr()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            None,
+        );
+
+        Self {
+            same_context,
+            a,
+            b,
+            sign_check_a,
+            sign_check_b,
+            lt_lo,
+            comparison_hi,
+            is_sgt,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _transaction: &Transaction<F>,
+        _call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let opcode = step.opcode.unwrap();
```

Comment:
Why is `opcode` an `Option` here? 
If you have a step there's no way on which it does not contain an opcode no?

Reply:
We could have internal `ExecutionState` such as `BeginTx` that doesn't have any corresponding opcode. That's why it's an option field in the `ExecStep`.

Reply:
Ohhh that's right!! Thanks!!! :)